core/pta/system.c

```diff
@@ -795,6 +795,138 @@ static TEE_Result system_dlsym(struct tee_ta_session *cs, uint32_t param_types,
 	return res;
 }
 
+static TEE_Result system_is_algo_supported(uint32_t param_types,
+					   TEE_Param params[TEE_NUM_PARAMS])
+{
+	uint32_t exp_pt = TEE_PARAM_TYPES(TEE_PARAM_TYPE_VALUE_INPUT,
+					  TEE_PARAM_TYPE_VALUE_OUTPUT,
+					  TEE_PARAM_TYPE_NONE,
+					  TEE_PARAM_TYPE_NONE);
+	TEE_Result st = TEE_ERROR_NOT_SUPPORTED;
+	uint32_t element = 0;
+	uint32_t alg = 0;
+
+	if (exp_pt != param_types)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	alg = params[0].value.a;
+	element = params[0].value.b;
+
+	switch (alg) {
```

Comment:
Please add ifdefs for `CFG_CRYPTO_AES` and all the others in `core/crypto.mk`

Reply:
I'm starting to wonder if this approach is the best. This huge switch case is going to take up quite a bit of memory.

How about only checking for algorithms which are optional in the spec and leave it to the `TEE_IsAlgorithmSupported()` function in the TA lib to return OK directly for all mandatory algorithms?

Reply:
Or perhaps everything can all be dealt with in `TEE_IsAlgorithmSupported()`.

Reply:
Ah, yes. As per the spec, they're supposed to be mandatory; but we do indeed have config flags so I will update the code.

Reply:
> Or perhaps everything can all be dealt with in `TEE_IsAlgorithmSupported()`.

The problem with this approach is: it is not very robust, since the TEE core is not queried. `libutee` is most often used as a static library so if the TA is not re-linked it may get wrong information.

Reply:
> This huge switch case is going to take up quite a bit of memory.

About 1KiB for the whole `system_is_algo_supported()` function (QEMU v7). Do you think it's too much?

Reply:
Yes, 1 KiB is a bit much. All of the algorithms here are mandatory? I thought that ECC was optional.

Reply:
You're right, ECC is optional and all others are mandatory.

Reply:
FYI, a non-debug QEMU build (`CFG_DEBUG_INFO=n`) is only ~0.5 KiB bigger in code size.

Reply:
That's an improvement, but I'd prefer a solution that probes using `utee_cryp_state_alloc()` and friends. We could consider changing those function a bit (returning `TEE_ERROR_NOT_SUPPORTED` early, before too many memory allocations) to make probing easier. That would be easier to maintain too.

Reply:
> That's an improvement, but I'd prefer a solution that probes using `utee_cryp_state_alloc()` and friends. We could consider changing those function a bit (returning `TEE_ERROR_NOT_SUPPORTED` early, before too many memory allocations) to make probing easier. 
> That would be easier to maintain too.

Easier said than done I'm afraid, because `{utee,syscall}_cryp_state_alloc()` do not take the same parameters as `TEE_IsAlgorithmSupported()`. And it can return success even with an invalid algorithm; invalid values are filtered out by `TEE_AllocateOperation()`. That is already a questionable choice since it's in libutee, not in the TEE core, so change the TEE binary without relinking TAs and everything breaks.

Given this last point (the user space libutee has to have the knowledge of which algorithms are supported or not), perhaps I should go back to my initial approach (not pushed here) and implement everything in `libutee`. This means no PTA, no additional code in the TEE core and just 0.5-1 KiB more of code in `libutee`. I could perhaps refactor the code a bit to share as much as possible between `TEE_AllocateOperation()` and `TEE_IsAlgorithmSupported()`.

Reply:
All algorithms but the asymmetric ciphers could be taken care of changing `syscall_cryp_state_alloc()` to allocate the context just after the `utc = to_user_ta_ctx(sess->ctx);` line.

The asymmetric algorithms are harder. Perhaps we could ifdef the `tee_cryp_obj_props[]` array to exclude currently unsupported key types and in that way be able to check if a key type is supported or not in the system PTA.

Reply:
> All algorithms but the asymmetric ciphers could be taken care of changing `syscall_cryp_state_alloc()` [...]

But what if `syscall_cryp_state_alloc()` returns an error different from "bad parameters" or "not supported"? "Out of memory" for instance? We could not draw any conclusion.

Reply:
We'd have to update `syscall_cryp_state_alloc()` so "not supported" is returned only and only if the algorithm isn't supported. I think it should be possible.